File: scripts/convert_xlsx_to_json.py
#!/usr/bin/env python3
"""
Simple XLSX -> JSON converter. Expects columns named (case-insensitive) one of:
- app_name (or name)
- package (or pkg, package_name)
- publisher (or publisher_name)

Writes a JSON array to the specified output path.
"""
import sys
import json
from pathlib import Path

def normalize_col(name):
    return name.strip().lower()

def find_col_map(headers):
    h = [normalize_col(x) for x in headers]
    mapping = {}
    for i, val in enumerate(h):
        if val in ('app_name', 'name', 'application'):
            mapping['app_name'] = i
        if val in ('package', 'pkg', 'package_name'):
            mapping['package'] = i
        if val in ('publisher', 'publisher_name', 'vendor'):
            mapping['publisher'] = i
    return mapping

def main():
    if len(sys.argv) < 3:
        print('Usage: convert_xlsx_to_json.py input.xlsx output.json')
        sys.exit(2)

    import pandas as pd

    inp = Path(sys.argv[1])
    out = Path(sys.argv[2])

    if not inp.exists():
        print('Input file not found:', inp)
        sys.exit(2)

    # read with pandas — it can handle xls/xlsx/csv
    df = pd.read_excel(inp, dtype=str)
    if df.empty:
        print('No rows found in', inp)
        sys.exit(1)

    # normalize columns
    cols = list(df.columns)
    colmap = find_col_map(cols)
    # attempt fallback heuristics
    if 'app_name' not in colmap and len(cols) >= 1:
        colmap['app_name'] = 0
    if 'package' not in colmap and len(cols) >= 2:
        colmap['package'] = 1
    if 'publisher' not in colmap and len(cols) >= 3:
        colmap['publisher'] = 2

    out_list = []
    for _, row in df.iterrows():
        try:
            name = str(row.iloc[colmap['app_name']]) if 'app_name' in colmap else ''
            pkg = str(row.iloc[colmap['package']]) if 'package' in colmap else ''
            pub = str(row.iloc[colmap['publisher']]) if 'publisher' in colmap else ''
        except Exception:
            continue
        if not name or not pkg:
            continue
        out_list.append({'app_name': name.strip(), 'package': pkg.strip(), 'publisher': pub.strip()})

    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(out_list, indent=2), encoding='utf8')
    print('Wrote', len(out_list), 'records to', out)

if __name__ == '__main__':
    main()

File: scripts/convert_xlsx_to_json.js
#!/usr/bin/env node
// Simple XLSX -> JSON converter using the 'xlsx' package.
const fs = require('fs');
const path = require('path');
const xlsx = require('xlsx');

function normalize(col){return String(col||'').trim().toLowerCase();}

function findHeaders(headers){
  const map = {};
  headers.forEach((h,i)=>{
    const n = normalize(h);
    if(['app_name','name','application'].includes(n)) map.app_name = i;
    if(['package','pkg','package_name'].includes(n)) map.package = i;
    if(['publisher','publisher_name','vendor'].includes(n)) map.publisher = i;
  });
  return map;
}

const argv = process.argv.slice(2);
if(argv.length < 2){
  console.error('Usage: node convert_xlsx_to_json.js input.xlsx output.json');
  process.exit(2);
}

const input = argv[0];
const output = argv[1];

if(!fs.existsSync(input)){ console.error('Input not found:', input); process.exit(2); }

const wb = xlsx.readFile(input, { cellDates: true });
const sheetName = wb.SheetNames[0];
const ws = wb.Sheets[sheetName];
const raw = xlsx.utils.sheet_to_json(ws, { header: 1 });
if(!raw || raw.length < 2){ console.error('No data found in sheet'); process.exit(1); }

const headers = raw[0];
const map = findHeaders(headers);
if(!map.app_name && headers.length>=1) map.app_name = 0;
if(!map.package && headers.length>=2) map.package = 1;
if(!map.publisher && headers.length>=3) map.publisher = 2;

const out = [];
for(let i=1;i<raw.length;i++){
  const row = raw[i];
  const name = String(row[map.app_name]||'').trim();
  const pkg = String(row[map.package]||'').trim();
  const pub = String(row[map.publisher]||'').trim();
  if(!name || !pkg) continue;
  out.push({ app_name: name, package: pkg, publisher: pub });
}

fs.mkdirSync(path.dirname(output), { recursive: true });
fs.writeFileSync(output, JSON.stringify(out, null, 2), 'utf8');
console.log('Wrote', out.length, 'records to', output);
